File: flink-scala-core/src/main/scala/com/atguigu/flink/day5/UpdateWindowResultLateEvent.scala
package com.atguigu.flink.day5

import java.time.Duration

import org.apache.flink.api.common.eventtime.{SerializableTimestampAssigner, WatermarkStrategy}
import org.apache.flink.api.common.state.{ValueState, ValueStateDescriptor}
import org.apache.flink.api.scala.typeutils.Types
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.ProcessFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.scala.function.ProcessWindowFunction
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.streaming.api.windowing.windows.TimeWindow
import org.apache.flink.util.Collector

object UpdateWindowResultLateEvent {

    def main(args: Array[String]): Unit = {
        val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
        env.setParallelism(1)
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)

        env.socketTextStream("hadoop102", 9999, '\n')
                .map(line => {
                    val data: Array[String] = line.split(" ")
                    (data(0), data(1).toLong * 1000L)
                })
                .assignTimestampsAndWatermarks(
                    WatermarkStrategy.forBoundedOutOfOrderness[(String, Long)](Duration.ofSeconds(5))
                            .withTimestampAssigner(new SerializableTimestampAssigner[(String, Long)] {
                                override def extractTimestamp(element: (String, Long), recordTimestamp: Long) = {
                                    element._2
                                }
                            })
                )
                .keyBy(_._1)
                .timeWindow(Time.seconds(5))
                .allowedLateness(Time.seconds(5))
                .process(new ProcessWindowFunction[(String, Long), String, String, TimeWindow] {

                    override def process(key: String, context: Context, elements: Iterable[(String, Long)], out: Collector[String]): Unit = {
                        val isUpdate: ValueState[Boolean] = context.windowState.getState(
                            new ValueStateDescriptor[Boolean]("is-update", Types.of[Boolean])
                        )
                        if (!isUpdate.value()) {
                            out.collect("first")
                            isUpdate.update(true)
                        } else {
                            out.collect("update result")
                        }
                    }
                })
                .print()


        env.execute()
    }

}

File: flink-java-core/src/main/java/com/atguigu/flink/day5/UpdateWindowResultLateEvent.java
package com.atguigu.flink.day5;

import org.apache.flink.api.common.eventtime.SerializableTimestampAssigner;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
import org.apache.flink.util.Collector;

import java.time.Duration;

/**
 * @author ：hmz
 * @date ：Created in 2020/9/15 16:53
 */
public class UpdateWindowResultLateEvent {

    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);

        env.socketTextStream("hadoop102", 9999)
                .map(new MapFunction<String, Tuple2<String, Long>>() {
                    @Override
                    public Tuple2<String, Long> map(String value) throws Exception {
                        String[] data = value.split(" ");
                        return Tuple2.of(data[0], Long.parseLong(data[1]) * 1000L);
                    }
                })
                .assignTimestampsAndWatermarks(
                        WatermarkStrategy
                                .<Tuple2<String, Long>>forBoundedOutOfOrderness(Duration.ofSeconds(5))
                                .withTimestampAssigner(
                                        new SerializableTimestampAssigner<Tuple2<String, Long>>() {
                                            @Override
                                            public long extractTimestamp(Tuple2<String, Long> element, long recordTimestamp) {
                                                return element.f1;
                                            }
                                        }
                                )
                )
                .keyBy(r -> r.f0)
                .timeWindow(Time.seconds(5))
                .allowedLateness(Time.seconds(5))
                .process(new ProcessWindowFunction<Tuple2<String, Long>, String, String, TimeWindow>() {

                    @Override
                    public void process(String s, Context context, Iterable<Tuple2<String, Long>> elements, Collector<String> out) throws Exception {
                        ValueState<Boolean> isUpdate = context.windowState().getState(
                                new ValueStateDescriptor<Boolean>("is-update", Types.BOOLEAN)
                        );
                        if (isUpdate.value() != null && isUpdate.value()) {
                            out.collect("update result");
                        } else {
                            out.collect("first");
                            isUpdate.update(true);
                        }
                    }
                })
                .print();

        env.execute();
    }

}
